chromadb/api/models/CollectionCommon.py

```diff
@@ -508,39 +536,58 @@ def _validate_and_prepare_upsert_request(
         documents: Optional[OneOrMany[Document]],
         images: Optional[OneOrMany[Image]],
         uris: Optional[OneOrMany[URI]],
-    ) -> Tuple[
-        IDs,
-        Embeddings,
-        Optional[Metadatas],
-        Optional[Documents],
-        Optional[URIs],
-    ]:
-        (
-            ids,
-            embeddings,
-            metadatas,
-            documents,
-            images,
-            uris,
-        ) = self._validate_embedding_set(
-            ids, embeddings, metadatas, documents, images, uris
+    ) -> RecordSet:
+        unpacked_record_set = self._unpack_record_set(
+            ids=ids,
+            embeddings=embeddings,
+            metadatas=metadatas,
+            documents=documents,
+            images=images,
+            uris=uris,
+        )
+
+        normalized_embeddings = (
+            self._normalize_embeddings(unpacked_record_set["embeddings"])
+            if unpacked_record_set["embeddings"] is not None
+            else None
         )
 
-        if embeddings is None:
-            if documents is not None:
-                embeddings = self._embed(input=documents)
-            else:
-                embeddings = self._embed(input=images)
+        self._validate_record_set(
+            ids=unpacked_record_set["ids"],
+            embeddings=normalized_embeddings,
+            metadatas=unpacked_record_set["metadatas"],
+            documents=unpacked_record_set["documents"],
+            images=unpacked_record_set["images"],
+            uris=unpacked_record_set["uris"],
+            require_embeddings_or_data=False,
+        )
+
+        prepared_embeddings = normalized_embeddings
+        if prepared_embeddings is None and does_record_set_contain_data(
```

Comment:
did this in the following pr because this `does_record_set_contain_data` takes in a record set. we refactored the code to use record set in the next pr

Reply:
We should move that refactor to this PR.

Reply:
done